File: scratch/websocket/python-ws-messaging/backend/messaging.py
# This is fast scratch and unpolished python source code, 
# which (only) intended to demonstrate websocket connectivity via tornado lib
# in most plain and straight forward way.
import tornado.websocket
import tornado.ioloop
import tornado.web

import os, binascii

import hashlib
import json
import time

class WSHandler(tornado.websocket.WebSocketHandler):
    '''
    Main handler of websocket

    INPUT/OUTPUT contracts:
    
    CLIENT: {msg: {to: '', txt: ''}}
    SERVER: {msg: {frm: '', private: '`true|fase`', txt: '`actual message`', ts: '`timestamp`'}}

    CLIENT: {users: ''}
    SERVER: {users: [[name, id], [name, id]]} (auto send when user in/out)

    SERVER: {err: 'error string'}
    SERVER: {connected: {id: '', name: ''}}
    '''
    users = dict()
    connections = dict()
    memory = dict(anon_count=0)
    def open(self):
        '''
        This will be called when a user try to connect to this server
        ws = new WebSocket('ws://192.168.72.129:8889/?name=your name'); ws.onmessage = function(resp) {...};
        '''
        if len(self.users) >= 20:
            self.write_message(dict(err='Unable to join server, it is full.'))
            self.close()
            return

        # prepare user's basic data
        name = self.get_query_argument('name', '')
        if not name:
            self.memory['anon_count'] = self.memory['anon_count'] + 1
            name = 'Anonymous #{0}'.format(self.memory['anon_count'])
        
        # save to memory for user's connection
        key = self._get_user_key()
        self.connections[key] = self
        self.users[key] = name
        
        # tell users their key and name saved
        reply = dict(connected=dict(id=key, name=name))
        self.write_message(json.dumps(reply))
        
        # tell other users someone joined
        self._notify_user_list_changes()
        print 'Total connections: {0}'.format(len(self.users))
        
    def on_message(self, message):
        '''
        React on message sent by client
        ws.send(JSON.stringify({...}))
        '''
        data = dict()
        
        try:
            data = json.loads(message)
        except:
            print 'Unexpected input, ignore: {0}'.format(message)
            import traceback
            print traceback.print_exc()
            self.write_message(dict(err='Unexpected input data.'))
            return
        
        for d in data:
            if d == 'msg':
                self._action_message(data[d]['txt'], data[d]['to'])
            elif d == 'users':
                self.write_message(self._msg_users_list())
            else:
                pass
                # no other yet

    def on_close(self):
        '''
        When a user disconnect from server
        '''
        key = False
        for c in self.connections:
            if self.connections[c] == self:
                key = c
                break
                
        if key != False:
            del self.connections[c]
            del self.users[c]
            
        self.close()
        
        # tell other users someone left
        self._notify_user_list_changes()
        print 'Connection closed. {0} connection(s) left.'.format(len(self.connections))

    def _get_user_key(self):
        '''
        Generate unique session key for each user
        '''
        key = ''
        try:
            base = '{0}{1}SALT'.format(self.request.remote_ip, self.request.headers['Sec-Websocket-Key'])
            key = hashlib.md5(base).hexdigest()
        except:
            print 'Failed to generate unique key for this request.'
            import traceback
            print traceback.print_exc()
        
        return key
    
    def _msg_users_list(self):
        '''
        return all connected clients
        '''
        users = dict(users=json.dumps(self.users))
        return json.dumps(users)
    
    def _notify_user_list_changes(self):
        '''
        notify all connected clients on client connected/disconnected
        '''
        for c in self.connections:
            self.connections[c].write_message(self._msg_users_list())
    
    def _action_message(self, message, target_user=None):
        '''
        Action: send message to client
        # OUT: {msg: {frm: '', private: '`true|fase`', txt: '`actual message`', ts: '`timestamp`'}}
        '''
        is_private_msg = True if target_user else False
        self_id = self._get_user_key()
        
        resp = dict(msg=dict(frm=self_id, private=is_private_msg, txt=message, ts=int(time.time())))
        
        if target_user:
            # to a user
            conn = self.connections[target_user]
            if conn:
                conn.write_message(resp)
        else:
            # to all            
            for c in self.connections:
                #if c == self_id:
                #    continue   
                self.connections[c].write_message(resp)

# setup routing
application = tornado.web.Application([
    (r'/msg', WSHandler)
])

# entry point
if __name__ == "__main__":
#    import daemon
#    log = open('tornado.' + str(port) + '.log', 'a+')
#    ctx = daemon.DaemonContext(stdout=log, stderr=log,  working_directory='.')
#    ctx.open()
    port = 8889
    application.listen(port)
    print 'Messaging service starting at port {0}'.format(port)
    tornado.ioloop.IOLoop.instance().start()

File: scratch/websocket/python-ws-messaging/web/javascript/messaging.js
/**
 * This is fast scratch and unpolished JavaScript source code, 
 * which (only) intended to demonstrate websocket connectivity in 
 * most plain and straight forward way.
 
 * INPUT/OUTPUT contracts:
 
 * CLIENT: {msg: {to: '', txt: ''}}
 * SERVER: {msg: {frm: '', private: '`true|fase`', txt: '`actual message`', ts: '`timestamp`'}}

 * CLIENT: {users: ''}
 * SERVER: {users: [[name, id], [name, id]]} (auto send when user in/out)

 * SERVER: {err: 'error string'}
 * SERVER: {connected: {id: '', name: ''}}
**/

var currentSessionId = null; // unique user id of current user
var allUsers = null; // a dict of users currently logged on
var showMessage = function(from, from_id, txt, ts, isSystemMessage) {
    // message received and show to user in dialog way
    var sys = isSystemMessage ? 'system' : '';  // need to distinguish system message and user message.
    var temp = $($.t('\
                <div class="message-container {from_id}"> \
                    <div class="from {sys}">{from}</div>: \
                    <div class="text {sys}" title="{ts}">{txt}</div> \
                </div>\
                ', {from_id: from_id, ts: ts.toLocaleString(), sys: sys}));
    $('.from', temp).text(from);
    $('.text', temp).text(txt);
    
    var conv = $('.conversation');
    conv.append(temp).scrollTop(conv[0].scrollHeight);
};

var msgHandle = function(action, args) {
    // handling instruction/message received
    if(action === 'users') {
        var data = JSON.parse(args);
        allUsers = data;
        var list = $('.user-list');
        list.children().remove();
        
        for(var user in data) {
            var name = data[user];
            if(user === currentSessionId) {
                name += ' (You)';
            }
            
            list.append($('<li>').prop('id', user).text(name));
        }
        
    } else if (action === 'msg') {
        
        var from = allUsers[args.frm],
            txt = args.txt,
            ts = new Date(parseInt(args.ts, 10) * 1000);
        
        if(args.frm === currentSessionId) {
            from += ' (You)';
        }
        showMessage(from, args.frm, txt, ts, false);
        
    } else if (action === 'err') {
        showMessage('[SYSTEM]', 'system', args, new Date(), true);
        
    } else if (action === 'connected') {
        currentSessionId = args.id;
    }
};

var wsHandle = {
    // handling communication with backend and dispatch valid message to `msgHandle'
    ws: null,  // websocket instance saved here
    bind: function() {
        wsHandle.ws.onmessage = wsHandle.onmessage;
        wsHandle.ws.onerror = wsHandle.onerror;
        wsHandle.ws.onclose = wsHandle.onclose;
        wsHandle.ws.onopen = wsHandle.onopen;
    },
    onmessage: function(resp) {
        var data = JSON.parse(resp.data);
        for(var d in data) {
            msgHandle(d, data[d]);
        }
    },
    
    onerror: function(resp) {
        console.log('Unexpected error', resp);
        var msg = 'Unexpected error: ' + resp;
        showMessage('[SYSTEM]', 'system', msg, new Date(), true);
    }, 
    
    onclose: function(resp) {
        $('.btn-send', '.messaging').prop('disabled', true);
        $('.user-list').children().remove();
        var connect_btn = $('.btn-connect', '.connect');
        connect_btn.attr('value', 'Connect');
        connect_btn.prop('disabled', false);
    },
    
    onopen: function(resp) {
        var connect_btn = $('.btn-connect', '.connect');
        connect_btn.attr('value', 'Disconnect');
        connect_btn.prop('disabled', false);
        $('.btn-send', '.messaging').prop('disabled', false);
    }
};

var clickHandle = {
    // Mapping functionalities and UI's clicks
    leftPanelBar: function() {
        var left = $(this),
            targetMargin = left.css('margin-left'),
            margin = parseFloat(targetMargin.replace('px', ''), 10);
        
        if(margin >= 0) {
            // auto get width percentage
            var marginPercentage = parseInt(left.css('width'), 10) / parseInt($('html').css('width'), 10);
            targetMargin = '-' + marginPercentage * 100 + '%';
            
        } else if (margin < 0) {
            targetMargin = '0px';
            
        } 
        
        left.animate({'margin-left': targetMargin}, 350);
    },
    
    connect_disconnect: function() {
        if($(this).val() === 'Connect') {
            clickHandle.connect($(this));
        } else {
            clickHandle.disconnect($(this));
        }
    },
    
    disconnect: function($btn) {
        if(wsHandle.ws) {
            wsHandle.ws.close();
        }
    },
    
    connect: function($btn) {
        //y = new WebSocket('ws://192.168.72.129:8889/msg?name=dsds')
        var $scope = $('.connect'),
            ip = $('.ip', $scope).val(),
            port = $('.port', $scope).val(),
            name = $('.name', $scope).val(),
            url = 'ws://' + ip + ':' + port + '/msg';
        
        if(name) {
            url += '?name=' + name;
        }
        
        if(wsHandle.ws) {
            wsHandle.ws.close();
            wsHandle.ws = null;
        }
        
        // here we connect to websocket, and instance are saved in `wsHandle`.`ws'
        wsHandle.ws = new WebSocket(url);
        wsHandle.bind();
        $btn.prop('disabled', true);
    },
    
    send: function() {
        var msg = $('.text', $('.messaging'));
        if(wsHandle.ws) {
            var obj = {msg: {to: '', txt: msg.val()}};
            wsHandle.ws.send(JSON.stringify(obj));
            msg.val('');
        }
    }
};

$(document).ready(function() {
    // entry point
    $('.left-panel').click(clickHandle.leftPanelBar);
    $('.btn-connect', $('.connect')).click(clickHandle.connect_disconnect);
    $('.btn-send', $('.messaging')).click(clickHandle.send);
});